pallene/to_ir.lua

```diff
@@ -209,66 +209,150 @@ function ToIR:convert_stat(cmds, stat)
+
+
+        if is_ipairs then
+            -- `ipairs` are desugared down to regular for-loops
+            -- ```
+            -- for i: T1, x: T2 in ipairs(xs) do
+            --   <loop body>
+            -- end
+            -- ```
+            -- would get compiled down to:
+            -- ```
+            -- local i_num: integer = 1
+            -- while true do
+            --   local x_dyn = xs[i_num]
+            --   if x_dyn == nil then
+            --     break
+            --   end
+            --   local i = i_num as T1
+            --   local x = x_dyn as T2 
+            --   <loop body>
+            --   i_num = i_num + 1
+            -- end
+            -- ```
+
+            -- the table passed as argument to `ipairs`
+            local arr =  exps[2].call_exp.args[1]
+            assert(arr._type.elem._tag == "types.T.Any")
+            local v_arr = ir.add_local(self.func, "$xs", arr._type)
+            self:exp_to_assignment(cmds, v_arr, arr)
+
+            -- local i_num: integer = 1
+            local v_inum = ir.add_local(self.func, "$"..decls[1].name.."_num", types.T.Integer())
+            local start = ir.Value.Integer(1)
+            table.insert(cmds, ir.Cmd.Move(stat.loc, v_inum, start))
+        
+            -- body of the while loop.
+            local body = {}
+
+            -- x_dyn = xs[i_num]
+            local v_x_dyn = ir.add_local(self.func, "$"..decls[2].name.."_dyn", types.T.Any())
+            local src_arr =  ir.Value.LocalVar(v_arr)
+            local src_i =  ir.Value.LocalVar(v_inum)
+            table.insert(body, ir.Cmd.GetArr(stat.loc, types.T.Any(), v_x_dyn, src_arr, src_i))
+
+            -- if x_dyn == nil then break end
+            local v_cond_checknil = ir.add_local(self.func, "$is_"..decls[2].name.."_nil", types.T.Boolean())
+            table.insert(body, ir.Cmd.IsNil(stat.loc, v_cond_checknil, ir.Value.LocalVar(v_x_dyn)))
+            table.insert(body, ir.Cmd.If(stat.loc, ir.Value.LocalVar(v_cond_checknil), ir.Cmd.Break(), ir.Cmd.Nop()))
+
+            -- local i: T1 = i_num as T1
+            local v_i = ir.add_local(self.func, decls[1].name, decls[1]._type)
+            self.loc_id_of_decl[decls[1]] = v_i
+            if decls[1]._type._tag == "types.T.Integer" then
+                table.insert(body, ir.Cmd.Move(stat.loc, v_i, ir.Value.LocalVar(v_inum)))
+            else
+                table.insert(body, ir.Cmd.ToDyn(stat.loc, types.T.Integer(), v_i, ir.Value.LocalVar(v_inum)))
+            end
+
+            -- local x = x_dyn as T2
+            local v_x = ir.add_local(self.func, decls[2].name, decls[2]._type)
+            self.loc_id_of_decl[decls[2]] = v_x
+            if decls[2]._type._tag == "types.T.Any" then
+                table.insert(body, ir.Cmd.Move(stat.loc, v_x, ir.Value.LocalVar(v_x_dyn)))
+            else
+                table.insert(body, ir.Cmd.FromDyn(stat.loc, decls[2]._type, v_x, ir.Value.LocalVar(v_x_dyn)))
+            end
```

Comment:
In this block of code we are also using `decls[2]`. If we choose to allow `for i in ipairs(xs)` then we need to update this part of the code. If we chose to forbid that kind of loop then this part of the code can stay unchanged.

Reply:
Ah yes. I don't think ipairs is used just for iterating from `1` to `#xs`, but at the same time It would be good to have Pallene resemble Lua as closely as possible and be a rough superset. So yea I'll change the `decls[2]` part.

Reply:
The golden rule that we follow is that if something is allowed in Pallene, then that program should behave the same as a the Lua program that you get by erasing all the type annotations. Pallene is allowed to give an error for things that would not be an error in Lua.

> Pallene mandates the LHS decl count and RHS return count be the same for loops so I'm kinda torn on this one.

If we choose to allow ipairs with a different number of variables, I think we would also have to change this rule accordingly (and the generic case of the for-in loops as well).

That said, since 99% of the time the programmers are going to want to use two variables in the ipairs, I think it's OK to keep the current behavior, at least for now. If it turns out that for-in loops with a mismatched number of arguments are something that people actually need we can always change it later.